unstructured_inference/models/chipper.py

```diff
@@ -457,6 +484,47 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
         )
 
 
+class NGramRepetitonStoppingCriteria(StoppingCriteria):
+    def __init__(self, repetition_window: int, skip_tokens: set = set()):
+        self.repetition_window = repetition_window
+        self.skip_tokens = skip_tokens
+
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
+        """
+        Args:
+            input_ids (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
+                Indices of input sequence tokens in the vocabulary.
+
+                Indices can be obtained using [`AutoTokenizer`]. See [`PreTrainedTokenizer.encode`]
+                and [`PreTrainedTokenizer.__call__`] for details.
+
+                [What are input IDs?](../glossary#input-ids)
+            scores (`torch.FloatTensor` of shape `(batch_size, config.vocab_size)`):
+                Prediction scores of a language modeling head. These can be scores for each
+                vocabulary token before SoftMax or scores for each vocabulary token after SoftMax.
+            kwargs (`Dict[str, Any]`, *optional*):
+                Additional stopping criteria specific kwargs.
+
+        Return:
+            `bool`. `False` indicates we should continue, `True` indicates we should stop.
+
+        """
+        num_batch_hypotheses = input_ids.shape[0]
+        cur_len = input_ids.shape[-1]
+
+        return any(
+            i
+            for la in _calc_banned_tokens(
+                input_ids,
+                num_batch_hypotheses,
+                self.repetition_window,
+                cur_len,
+            )
+            for i in la
+            if i not in self.skip_tokens
```

Comment:
this is not easy to read, why not just iterate through nested for loops the old-fashioned way?

Reply:
The loop looks like this now.

        for banned_tokens in _calc_banned_tokens(
            input_ids,
            num_batch_hypotheses,
            self.repetition_window,
            cur_len,
        ):
            for token in banned_tokens:
                if token not in self.skip_tokens:
                    return True

        return False

Reply:
thanks!